### PR TITLE
More fixes to gossip

### PR DESCRIPTION
**Pull request by freesig**

### CODE REVIEW

File: crates/holochain_sqlite/src/sql/cell/schema.sql [hunk 1]
```diff
@@ -92,6 +92,7 @@ CREATE TABLE IF NOT EXISTS DhtOp (
 
     storage_center_loc          INTEGER   NOT NULL,
     authored_timestamp_ms       INTEGER   NOT NULL,
+    private_entry               INTEGER   NULL,  -- BOOLEAN
```
Comment (neonphog) [inline crates/holochain_sqlite/src/sql/cell/schema.sql:95]: is the query faster checking NULL vs false? If so, totally cool with this, otherwise it might be more semantically readable like:
```suggestion
    private_entry               INTEGER   DEFAULT FALSE NOT NULL,  -- BOOLEAN
```

**Review by neonphog — APPROVED** [review]

File: crates/holochain_state/src/source_chain.rs [hunk 2]
```diff
@@ -680,7 +680,11 @@ impl SourceChain {
                     // If this is a countersigning session we want to withhold
                     // publishing the ops until the session is successful.
                     if is_countersigning_session {
-                        set_withhold_publish(txn, op_hash)?;
+                        set_withhold_publish(txn, op_hash.clone())?;
+                    }
+
+                    if is_private_entry {
+                        set_private_entry_store_entry(txn, op_hash)?;
                     }
```
Comment (maackle) [inline crates/holochain_state/src/source_chain.rs:688]: This goes beyond the scope of this PR but I don't think it's great to just keep performing extra queries to modify the newly created DB rows when the correct values could just have just been set in the original INSERT query, is there a reason we do it this way?

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/store.rs [hunk 3]
```diff
@@ -209,3 +212,25 @@ pub(super) async fn put_ops(
 
     Ok(())
 }
+
+pub(super) async fn put_ops_direct(
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/store.rs:216]:
```suggestion
/// Specifies exactly which ops should go to which agents, in contrast to
/// `put_ops` which first figures out which ops should be held by which agents.
pub(super) async fn put_ops_direct(
```

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs [hunk 4]
```diff
@@ -560,30 +560,54 @@ impl ShardedGossipLocal {
...
             &self.space,
             local_agents.iter(),
-            op_hashes,
+            ops_needed.into_iter().collect(),
+            true,
         )
         .await?
         .into_iter()
         .collect();
+        store::put_ops_direct(&self.evt_sender, &self.space, needed_op_hashes, ops).await?;
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:609]: Why wasn't `put_ops` sufficient here? I feel uneasy about doing the arc logic outside of this function and then passing the calculated results directly in. It seems like `put_ops` should be doing this. Why is this needed and do we just need to modify `put_ops` to have logic more like what's above?

**Review by maackle — CHANGES_REQUESTED** [review]
Looks good but I have a question about `put_ops_direct`

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs (hunk 4 above)
Comment (freesig) [reply crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:609]: Because we put_ops combines the results so you have no idea who has which hashes. We need to know which agent has which hashes so we can not send ops to agents that already have them.

File: crates/holochain_state/src/source_chain.rs (hunk 2 above)
Comment (freesig) [reply crates/holochain_state/src/source_chain.rs:688]: Because this information comes from the header. There's nothing in the dhtop light to determine if it has a private entry.

Comment (maackle) [reply crates/holochain_state/src/source_chain.rs:688]: Yes, but we have the header at the time we're inserting the op, so we could insert the op and set the appropriate fields at the same time as part of the insert, no?

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs (hunk 4 above)
Comment (maackle) [reply crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:609]: @freesig then my next question is, is there any reason why we wouldn't always just do it this way? Do we actually want to use `put_ops` in other situations or is it more optimal to just do it this way, removing extraneous puts? I'm questioning whether we can combine these two methods.

File: crates/holochain_state/src/source_chain.rs (hunk 2 above)
Comment (freesig) [reply crates/holochain_state/src/source_chain.rs:688]: Yeh I guess we could with this. A few of the others calls are specific to authoring ops vs inserting them when you get them through gossip / publish however setting the private entry flag shouldn't matter where it comes from as we shouldn't be gossiping private store entries anyway. I'll make the change to this.

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs (hunk 4 above)
Comment (freesig) [reply crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:609]: I'm not sure about this. If we move in the direction I was talking to about today (only having one database) then we always know that the ops are for that one database and don't need to check. It's expensive to pull down all the hashes an agent already has to check if we should send up the ops. This is part of why local sync is expensive. This is also a really good example of how we could simplify gossip if we had a single database.

I think I will leave this with a comment that we might be removing local sync soon and this `put_ops_direct` function soon. Unless you feel I should just remove it from this PR?

Comment (maackle) [reply crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:609]: > If we move in the direction I was talking to about today (only having one database) then we always know that the ops are for that one database and don't need to check

Isn't the point of `put_ops` that it does this extra check? So wouldn't that be an argument in favor of just removing `put_ops`?

**Comment by freesig** [comment]
Closing in favor of #978

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs (hunk 4 above)
Comment (freesig) [reply crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:609]: Because if you have a single database you will never get sent ops you don't already have due to the bloom filter
